File: src/jsMain/kotlin/com/tfjs/ModelManager.kt
package com.tfjs

import kotlinx.coroutines.*
import kotlin.js.Promise

/**
 * External declarations for TensorFlow.js
 */
@JsModule("@tensorflow/tfjs")
@JsNonModule
external object tf {
    fun loadLayersModel(path: String): Promise<dynamic>
    fun tensor2d(data: Array<Array<Double>>): dynamic
    fun tidy(fn: () -> dynamic): dynamic
}

/**
 * Manages TensorFlow.js model loading and predictions
 */
class ModelManager {
    private var model: dynamic = null
    private val scope = CoroutineScope(Dispatchers.Default)
    
    /**
     * Load the TensorFlow.js model
     */
    fun loadModel(callback: (Boolean) -> Unit) {
        scope.launch {
            try {
                console.log("Loading model from: model/model.json")
                model = tf.loadLayersModel("model/model.json").await()
                console.log("Model loaded successfully")
                console.log("Model details:", model)
                callback(true)
            } catch (e: Throwable) {
                console.error("Error loading model:", e.message ?: "Unknown error")
                console.error("Stack trace:", e)
                callback(false)
            }
        }
    }
    
    /**
     * Make a prediction using the loaded model
     * @param input Array of 4 double values
     * @return Promise with FloatArray of predictions
     */
    fun predict(input: Array<Double>): Promise<FloatArray>? {
        if (model == null) {
            console.error("Prediction attempted but model is not loaded!")
            return null
        }
        
        console.log("Making prediction with input:", input.joinToString())
        
        return Promise { resolve, reject ->
            try {
                val result = tf.tidy {
                    // Create 2D tensor with shape [1, 4]
                    val inputTensor = tf.tensor2d(arrayOf(input))
                    console.log("Input tensor created")
                    
                    // Make prediction
                    val prediction = model.predict(inputTensor)
                    console.log("Prediction made")
                    
                    // Extract data
                    prediction.data()
                }
                
                result.then { data ->
                    val floatArray = data as FloatArray
                    console.log("Prediction result:", floatArray.joinToString())
                    resolve(floatArray)
                }.catch { error ->
                    console.error("Prediction error:", error)
                    reject(error)
                }
            } catch (e: Throwable) {
                console.error("Exception during prediction:", e)
                reject(e)
            }
        }
    }
    
    /**
     * Check if model is loaded
     */
    fun isModelLoaded(): Boolean {
        return model != null
    }
    
    /**
     * Get model information
     */
    fun getModelInfo(): String {
        return if (model != null) {
            "Model loaded and ready"
        } else {
            "Model not loaded"
        }
    }
}

File: src/jsMain/typescript/core/ModelManager.ts
/**
 * ModelManager - Handles TensorFlow.js model loading and predictions
 */

import * as tf from '@tensorflow/tfjs';
import type { 
    ModelConfig, 
    PredictionResult, 
    ModelLoadCallback, 
    PredictionCallback 
} from '../types/tensorflow';

export class ModelManager {
    private model: tf.LayersModel | null = null;
    private config: ModelConfig;
    private isLoading: boolean = false;

    constructor(config: ModelConfig) {
        this.config = config;
    }

    /**
     * Load the TensorFlow.js model
     */
    public async loadModel(callback: ModelLoadCallback): Promise<void> {
        if (this.isLoading) {
            console.warn('Model is already being loaded');
            return;
        }

        this.isLoading = true;
        console.log(`🔄 Loading model from: ${this.config.modelPath}`);

        try {
            this.model = await tf.loadLayersModel(this.config.modelPath);
            console.log('✅ Model loaded successfully');
            console.log('📊 Model details:', {
                inputs: this.model.inputs.map(i => i.shape),
                outputs: this.model.outputs.map(o => o.shape),
            });
            
            this.isLoading = false;
            callback(true);
        } catch (error) {
            console.error('❌ Error loading model:', error);
            this.isLoading = false;
            callback(false, error as Error);
        }
    }

    /**
     * Make a prediction using the loaded model
     */
    public async predict(input: number[]): Promise<PredictionResult | null> {
        if (!this.model) {
            console.error('⚠️ Model not loaded!');
            return null;
        }

        if (input.length !== this.config.inputShape[0]) {
            console.error(
                `❌ Invalid input size. Expected ${this.config.inputShape[0]}, got ${input.length}`
            );
            return null;
        }

        console.log('🔮 Making prediction with input:', input);

        return tf.tidy(() => {
            // Create input tensor with shape [1, inputSize]
            const inputTensor = tf.tensor2d([input], [1, this.config.inputShape[0]]);
            
            // Make prediction
            const prediction = this.model!.predict(inputTensor) as tf.Tensor;
            
            // Get probabilities
            const probabilities = prediction.dataSync() as Float32Array;
            
            // Find predicted class
            let maxIndex = 0;
            let maxValue = probabilities[0];
            
            for (let i = 1; i < probabilities.length; i++) {
                if (probabilities[i] > maxValue) {
                    maxValue = probabilities[i];
                    maxIndex = i;
                }
            }

            const result: PredictionResult = {
                predictedClass: maxIndex,
                confidence: maxValue,
                probabilities: probabilities
            };

            console.log('✅ Prediction result:', result);
            return result;
        });
    }

    /**
     * Check if model is loaded
     */
    public isModelLoaded(): boolean {
        return this.model !== null;
    }

    /**
     * Get model information
     */
    public getModelInfo(): string {
        if (!this.model) {
            return 'Model not loaded';
        }

        const inputShape = this.model.inputs[0].shape;
        const outputShape = this.model.outputs[0].shape;
        
        return `Model loaded - Input: ${inputShape}, Output: ${outputShape}`;
    }

    /**
     * Dispose of the model and free memory
     */
    public dispose(): void {
        if (this.model) {
            this.model.dispose();
            this.model = null;
            console.log('🗑️ Model disposed');
        }
    }
}